File: server/datasets/java_source-code/Multiple-G2/structural_26_avl_tree.java
class Node_1 {
    int k_1, h_1;
    Node_1 left, right;
    Node_1(int d) {
        k_1 = d;
        h_1 = 1;
    }
}
public class Tree_1 {
    public static void main(String[] args) {
        Tree_1 t = new Tree_1();
        Node_1 r = null;
        r = t.ins_1(r, 10);
    }
    int max(int a, int b) {
        return (a > b) ? a : b;
    }
    int h_1(Node_1 N) {
        if (N == null) return 0;
        return N.h_1;
    }
    int gb_1(Node_1 N) {
        if (N == null) return 0;
        return h_1(N.left) - h_1(N.right);
    }
    Node_1 upd_1(Node_1 n_1) {
        n_1.h_1 = 1 + max(h_1(n_1.left), h_1(n_1.right));
        return n_1;
    }
    Node_1 rr_1(Node_1 y) {
        Node_1 x = y.left;
        Node_1 T2 = x.right;
        x.right = y;
        y.left = T2;
        upd_1(y);
        upd_1(x);
        return x;
    }
    Node_1 rl_1(Node_1 x) {
        Node_1 y = x.right;
        Node_1 T2 = y.left;
        y.left = x;
        x.right = T2;
        upd_1(x);
        upd_1(y);
        return y;
    }
    Node_1 ins_1(Node_1 n_1, int k_1) {
        if (n_1 == null) return (new Node_1(k_1));
        if (k_1 < n_1.k_1) n_1.left = ins_1(n_1.left, k_1);
        else if (k_1 > n_1.k_1) n_1.right = ins_1(n_1.right, k_1);
        else return n_1;
        upd_1(n_1);
        int b_1 = gb_1(n_1);
        if (b_1 > 1 && k_1 < n_1.left.k_1) return rr_1(n_1);
        if (b_1 < -1 && k_1 > n_1.right.k_1) return rl_1(n_1);
        if (b_1 > 1 && k_1 > n_1.left.k_1) {
            n_1.left = rl_1(n_1.left);
            return rr_1(n_1);
        }
        if (b_1 < -1 && k_1 < n_1.right.k_1) {
            n_1.right = rr_1(n_1.right);
            return rl_1(n_1);
        }
        return n_1;
    }
}

File: server/datasets/python_source-code/Multiple-G2/structural_26_avl_tree.py
class Node_1:
    def __init__(self, v_1):
        self.v_1 = v_1
        self.r_1 = None
        self.l_1 = None
        self.height = 1

def h_1(node):
    if not node:
        return 0
    return node.height

def b_1(node):
    if not node:
        return 0
    return h_1(node.l_1) - h_1(node.r_1)

def rr_1(y):
    x = y.l_1
    sub = x.r_1
    x.r_1 = y
    y.l_1 = sub
    y.height = 1 + max(h_1(y.l_1), h_1(y.r_1))
    x.height = 1 + max(h_1(x.l_1), h_1(x.r_1))
    return x

def rl_1(x):
    y = x.r_1
    sub = y.l_1
    y.l_1 = x
    x.r_1 = sub
    x.height = 1 + max(h_1(x.l_1), h_1(x.r_1))
    y.height = 1 + max(h_1(y.l_1), h_1(y.r_1))
    return y

def ins_1(rt_1, v_1):
    if not rt_1:
        return Node_1(v_1)
    if v_1 < rt_1.v_1:
        rt_1.l_1 = ins_1(rt_1.l_1, v_1)
    elif v_1 > rt_1.v_1:
        rt_1.r_1 = ins_1(rt_1.r_1, v_1)
    else:
        return rt_1

    rt_1.height = 1 + max(h_1(rt_1.l_1), h_1(rt_1.r_1))
    bf_1 = b_1(rt_1)

    if bf_1 > 1 and v_1 < rt_1.l_1.v_1:
        return rr_1(rt_1)
    if bf_1 < -1 and v_1 > rt_1.r_1.v_1:
        return rl_1(rt_1)
    if bf_1 > 1 and v_1 > rt_1.l_1.v_1:
        rt_1.l_1 = rl_1(rt_1.l_1)
        return rr_1(rt_1)
    if bf_1 < -1 and v_1 < rt_1.r_1.v_1:
        rt_1.r_1 = rr_1(rt_1.r_1)
        return rl_1(rt_1)

    return rt_1
